> Nama : Rizky Ramdhani Koswara

> NPM : 11122300

> Kelas : 4KA25

# 1. Import Library

In [ ]:
import tensorflow as tf

In [ ]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Embedding, LSTM, Dense, Bidirectional, Dropout
from tensorflow.keras.models import Sequential
from tensorflow.keras import regularizers
from tensorflow.keras.optimizers import Adam
import numpy as np

### **Penjelasan Logika & Output: 1. Import Library**

**Logika Kode:**
Blok ini ngeberesin semua impor *library* yang dibutuhin.
* `tensorflow` diimpor sebagai `tf`, ini *framework* utamanya.
* `Tokenizer` dipake buat ngubah teks (kata-kata) jadi angka (token) biar bisa diproses model.
* `pad_sequences` dipake buat bikin semua urutan teks jadi sama panjang.
* `Embedding`, `LSTM`, `Dense`, `Bidirectional`, `Dropout` adalah lapisan-lapisan (layers) yang bakal kita pake buat ngebangun arsitektur model RNN kita. `LSTM` itu inti modelnya, `Bidirectional` bikin LST-nya bisa baca teks dari depan-belakang dan belakang-depan.
* `Sequential` adalah cara kita ngebangun modelnya, lapisan demi lapisan.
* `regularizers` dipake buat nambahin penalti ke *loss* biar model gak *overfitting*.
* `Adam` itu *optimizer* yang kita pake buat ngelatih model.
* `numpy` dipake buat ngurusin *array* angka.

**Output/Efek:**
Gak ada *output* visual. Semua *library* dan modul yang dibutuhin jadi siap dipake di memori.

# 2. Mengunduh dataset shakesphere

In [ ]:
!wget --no-check-certificate \
  https://huggingface.co/arnavmahapatra/gpt2-sonnet-generators/blob/main/shakespeare.txt \
  -O sonnets.txt

--2024-11-15 08:29:47--  https://huggingface.co/arnavmahapatra/gpt2-sonnet-generators/blob/main/shakespeare.txt
Resolving huggingface.co (huggingface.co)... 18.238.109.102, 18.238.109.92, 18.238.109.52, ...
Connecting to huggingface.co (huggingface.co)|18.238.109.102|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1094068 (1.0M) [text/html]
Saving to: ‘sonnets.txt’

sonnets.txt         100%[===================>]   1.04M  4.38MB/s    in 0.2s    

2024-11-15 08:29:47 (4.38 MB/s) - ‘sonnets.txt’ saved [1094068/1094068]



### **Penjelasan Logika & Output: 2. Mengunduh dataset shakesphere**

**Logika Kode:**
Perintah `!wget` ini adalah perintah *shell* (bukan Python) buat ngunduh file dari internet.
* Dia ngambil file `shakespeare.txt` dari alamat URL di Hugging Face.
* `--no-check-certificate` dipake buat nge-*skip* verifikasi sertifikat SSL, jaga-jaga kalo ada masalah.
* `-O sonnets.txt` nyuruh `wget` buat nyimpen filenya dengan nama `sonnets.txt` di direktori kerja Colab kita.

**Output/Efek:**
Outputnya nunjukkin proses *download* file. Kalo sukses, bakal ada tulisan kayak "Saving to: ‘sonnets.txt’" dan "‘sonnets.txt’ saved". Artinya, kita sekarang punya file teks `sonnets.txt` yang isinya karya-karya Shakespeare.

# 3. Mendefinisikan Tokenizer dan Menyiapkan Training Data

Langkah  selanjutnya  yaitu  melakukan  proses  tokenisasi  dan  menyiapkan  data  yang  akan dilatih.  tokenisasi  adalah  proses  untuk  membagi  teks  yang  dapat  berupa  kalimat,  paragraf atau dokumen, menjadi token-token/bagian-bagian tertentu. Pada proses ini suatu kata yang ada pada  data sheet akan disimbolkan dengan angka secara acak sampai jumlah  kata yang ada  pada  datasheet  tersebut.  Pada  datasheet  ini  jumlah  katanya  yaitu  sampai  2900. Contohnya yaitu angka 1 didefinisikan untuk kata “and”, 2 untuk kata “the” dan seterusnya. Berikut adalah kodingan serta output dari proses tokenizer.

In [ ]:
tokenizer = Tokenizer()

data = open('sonnets.txt').read()

corpus = data.lower().split('\n')

tokenizer.fit_on_texts(corpus)
total_words = len(tokenizer.word_index) + 1

print(tokenizer.word_index)
print(total_words)

{'and': 1, 'the': 2, 'to': 3, 'of': 4, 'my': 5, 'i': 6, 'in': 7, 'that': 8, 'thy': 9, 'thou': 10, 'with': 11, 'for': 12, 'a': 13, 'but': 14, 'love': 15, 'is': 16, 'thee': 17, 'not': 18, 'me': 19, 'so': 20, 'you': 21, 'all': 22, 'be': 23, 'his': 24, 'when': 25, 'as': 26, 'which': 27, 'your': 28, 'this': 29, 'it': 30, 'doth': 31, 'by': 32, 'or': 33, 'do': 34, 'from': 35, 'then': 36, 'on': 37, 'no': 38, 'are': 39, 'their': 40, 'shall': 41, 'what': 42, 'have': 43, 'mine': 44, 'time': 45, 'if': 46, 'sweet': 47, 'more': 48, 'beauty': 49, 'art': 50, 'than': 51, 'where': 52, 'nor': 53, 'can': 54, 'should': 55, 'will': 56, 'now': 57, 'o': 58, 'how': 59, 'they': 60, 'he': 61, 'yet': 62, 'hath': 63, 'eyes': 64, 'make': 65, 'thine': 66, 'fair': 67, 'eye': 68, 'him': 69, 'one': 70, 'still': 71, 'like': 72, 'being': 73, 'every': 74, 'own': 75, 'praise': 76, 'some': 77, 'true': 78, 'such': 79, 'live': 80, 'heart': 81, 'see': 82, 'though': 83, 'day': 84, 'world': 85, 'were': 86, 'those': 87, 'give': 8

### **Penjelasan Logika & Output: 3. Mendefinisikan Tokenizer dan Menyiapkan Training Data**

**Logika Kode:**
Ini adalah langkah krusial buat nyiapin data teks mentah jadi data *training* yang siap dipake model.
1.  `tokenizer = Tokenizer()`: Bikin objek *tokenizer* baru.
2.  `data = open('sonnets.txt').read()`: Ngebaca semua isi file `sonnets.txt` ke dalam satu *string* gede.
3.  `corpus = data.lower().split('\n')`: Ngebersihin data. `lower()` ngubah semua teks jadi huruf kecil (biar "The" dan "the" dianggap sama). `split('\n')` mecah teks jadi daftar (list) berdasarkan baris baru. Jadi, `corpus` sekarang isinya *list* dari tiap baris puisi.
4.  `tokenizer.fit_on_texts(corpus)`: Ini "ngelatih" *tokenizer*-nya. Dia ngebaca semua baris di `corpus` dan bikin "kamus" (`word_index`) dari semua kata unik yang dia temuin.
5.  `total_words = len(tokenizer.word_index) + 1`: Ngitung jumlah total kata unik. Ditambah 1 buat nge-handle *padding* (token 0) nanti.
6.  `print(tokenizer.word_index)`: Nampilin kamus yang udah dibuat.
7.  `print(total_words)`: Nampilin jumlah total kata unik + 1.

**Output/Efek:**
* Output pertama adalah `word_index`, yaitu kamus kata-ke-angka. Contohnya `{'and': 1, 'the': 2, 'to': 3, ...}`. Ini nunjukkin kata yang paling sering muncul dapet angka (indeks) kecil.
* Output kedua adalah jumlah total kata unik plus satu, di contoh ini 2900.

In [ ]:
input_sequences = []

for line in corpus:
  token_list = tokenizer.texts_to_sequences ([line])[0]
  #print("LIST"+str(token_list))
  for i in range(1, len(token_list)):
    n_gram_sequence = token_list[:i+1]
    input_sequences.append(n_gram_sequence)


##pad sequences
max_sequence_len = max([len(seq) for seq in input_sequences])

print(max_sequence_len, total_words)

input_sequences = np.array(pad_sequences(input_sequences, padding='pre', maxlen=max_sequence_len))

#create predictors and labels
xs, labels = input_sequences[:,:-1], input_sequences[:,-1]

ys = tf.keras.utils.to_categorical (labels, num_classes=total_words)

11 2900


### **Penjelasan Logika & Output: (Lanjutan) Penyiapan Data Sequence**

**Logika Kode:**
Setelah punya kamus, kita harus ubah teks jadi urutan angka buat *training*. Tujuannya adalah *word prediction*: dari beberapa kata sebelumnya, tebak satu kata berikutnya.
1.  `input_sequences = []`: Bikin *list* kosong buat nyimpen semua urutan data.
2.  `for line in corpus:`: Nge-*loop* tiap baris puisi di `corpus`.
3.  `token_list = tokenizer.texts_to_sequences([line])[0]`: Ngubah satu baris puisi (misal: "from fairest creatures") jadi urutan angka (misal: `[35, 714, 1160]`).
4.  `for i in range(1, len(token_list)):`: Ini bagian penting. Dia bikin *n-gram* atau sub-kalimat. Kalo urutannya `[35, 714, 1160]`, dia bakal bikin:
    * `[35, 714]` (input: "from", target: "fairest")
    * `[35, 714, 1160]` (input: "from fairest", target: "creatures")
5.  `input_sequences.append(n_gram_sequence)`: Masukin semua sub-kalimat tadi ke `input_sequences`.
6.  `max_sequence_len = max([...])`: Nyari tau urutan terpanjang dari semua sub-kalimat yang kita buat.
7.  `input_sequences = np.array(pad_sequences(...))`: Bikin semua urutan di `input_sequences` jadi sama panjang (sepanjang `max_sequence_len`). `padding='pre'` nambahin angka 0 di *depan* urutan yang lebih pendek.
8.  `xs, labels = ...`: Mecah data. `xs` (input) adalah semua token *kecuali* yang terakhir. `labels` (target) adalah *cuma* token terakhir.
9.  `ys = tf.keras.utils.to_categorical(...)`: Ngubah `labels` (yang cuma satu angka, misal `1160`) jadi format *one-hot encoding*. Jadi *array* sepanjang `total_words` (2900) yang isinya 0 semua kecuali di indeks ke-1160 yang nilainya 1. Ini format yang dibutuhin buat *loss* `categorical_crossentropy`.

**Output/Efek:**
* Nge-print `max_sequence_len` (11) dan `total_words` (2900).
* Variabel `xs` dan `ys` sekarang siap dipake buat *training*. `xs` bentuknya (jumlah_urutan, 10) dan `ys` bentuknya (jumlah_urutan, 2900).

# 4. Mendefinisikan Arsitektur Model

Langkah selanjutnya yaitu mendefinisikan arsitektur , pada tahap ini kita menentukan berapa total kata yang ada pada artikel tersebut kemudian menentukan urutan setiap kata yang ada pada  artikel  tersebut.  Berikut  adalah  kodingan  serta  output  dari  mendefinisikan  arsitektur modelnya.

In [ ]:
model = Sequential()
model.add(Embedding(total_words, 100, input_length=max_sequence_len-1))
model.add(Bidirectional(LSTM(150, return_sequences=True)))
model.add(Dropout(0.3))
model.add(Bidirectional(LSTM(96)))
model.add(Dense(total_words//2, activation='relu', kernel_regularizer=regularizers.l2(0.01)))
model.add(Dense(total_words, activation='softmax'))

print(model.summary())


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential_6"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_6 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ bidirectional_12 (Bidirectional)     │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_6 (Dropout)                  │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ bidirectional_13 (Bidirectional)     │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_12 (Dense)                     │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_13 (Dense)                     │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

None


### **Penjelasan Logika & Output: 4. Mendefinisikan Arsitektur Model**

**Logika Kode:**
Ini bagian ngebangun arsitektur model LSTM-nya pake `Sequential`.
1.  `Embedding(total_words, 100, ...)`: *Layer* pertama. Ini ngubah tiap kata (angka token) di input `xs` jadi vektor 100 dimensi. Ini cara model belajar "arti" dan hubungan antar kata.
2.  `Bidirectional(LSTM(150, return_sequences=True))`: *Layer* LSTM bolak-balik (Bidirectional) dengan 150 unit. `return_sequences=True` artinya dia ngeluarin *output* buat tiap token di urutan, bukan cuma *output* terakhir. Ini perlu karena kita mau numpuk *layer* LSTM lagi.
3.  `Dropout(0.3)`: *Layer* *dropout* buat ngebuang 30% koneksi neuron secara acak pas *training*. Ini buat ngurangin *overfitting*.
4.  `Bidirectional(LSTM(96))`: *Layer* LSTM bolak-balik kedua. Yang ini `return_sequences`-nya *default* (False), jadi dia cuma ngeluarin satu *output* di akhir urutan.
5.  `Dense(total_words//2, ...)`: *Layer* `Dense` biasa. `kernel_regularizer=regularizers.l2(0.01)` nambahin L2 *regularization* buat bantu cegah *overfitting* lagi.
6.  `Dense(total_words, activation='softmax')`: *Layer* *output* terakhir. Harus punya neuron sebanyak `total_words` (2900). Aktivasi `softmax` bakal ngasih probabilitas buat tiap kata di kamus, nunjukkin kata mana yang paling mungkin jadi kata berikutnya.
7.  `print(model.summary())`: Nampilin ringkasan arsitektur model.

**Output/Efek:**
* Outputnya adalah ringkasan model (`model.summary()`). Harusnya nunjukkin daftar *layer*, bentuk *output*, dan jumlah parameter. *Catatan: Output di notebook lo nunjukkin "unbuilt" karena modelnya belum dikasih data, tapi setelah di-compile dan di-fit, summary-nya bakal bener.*

# 5. Training Data

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
history = model.fit(xs, ys, epochs=100, verbose=1)

Epoch 1/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 62s 143ms/step - accuracy: 0.0237 - loss: 7.7045
Epoch 2/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 55s 145ms/step - accuracy: 0.0209 - loss: 6.4439
Epoch 3/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 80s 139ms/step - accuracy: 0.0265 - loss: 6.3765
Epoch 4/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 82s 138ms/step - accuracy: 0.0268 - loss: 6.3082
Epoch 5/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 82s 137ms/step - accuracy: 0.0316 - loss: 6.2169
Epoch 6/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 83s 139ms/step - accuracy: 0.0419 - loss: 6.0480
Epoch 7/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 83s 142ms/step - accuracy: 0.0393 - loss: 5.9812
Epoch 8/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 83s 147ms/step - accuracy: 0.0433 - loss: 5.9233
Epoch 9/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 78s 137ms/step - accuracy: 0.0506 - loss: 5.7860
Epoch 10/10
380/380 ━━━━━━━━━━━━━━━━━━━━ 81s 135ms/step - accuracy: 0.0470 - loss: 5.7064


### **Penjelasan Logika & Output: 5. Training Data**

**Logika Kode:**
Ini proses ngelatih model yang udah kita bikin.
1.  `model.compile(...)`: Nyatuin model. Kita tentuin *loss function*-nya (`categorical_crossentropy`, cocok buat klasifikasi multi-kelas pake *softmax*), *optimizer*-nya (`adam`), dan metrik yang mau diliat (`accuracy`).
2.  `history = model.fit(xs, ys, epochs=100, verbose=1)`: Ini dia perintah buat mulai *training*.
    * `xs`: Data input (urutan kata).
    * `ys`: Data target (kata berikutnya, udah di-*one-hot encode*).
    * `epochs=100`: Ngulang proses *training* di seluruh dataset sebanyak 100 kali. Di *notebook* lo, ini di-*run* cuma 10 *epoch*.
    * `verbose=1`: Nampilin *progress bar* dan metrik buat tiap *epoch*.

**Output/Efek:**
* Outputnya adalah *log* proses *training*. Buat tiap *epoch* (dari 10), dia bakal nunjukkin *progress bar*, waktu yang dibutuhin, nilai `loss`, dan `accuracy`.
* Contoh: `Epoch 1/10 380/380 [...] - accuracy: 0.0237 - loss: 7.7045`.
* Akurasi awalnya pasti kecil banget karena nebak 1 dari 2900 kata itu susah. Seiring *epoch* berjalan, akurasinya harusnya naik dan *loss*-nya turun.

# 6. Membuat Perintah Untuk 100 Kata Selanjutnya

Langkah selanjutnya yaitu membuat perintah untuk RNN supaya dapat memprediksi 100kata selanjutnya pada artikel yang ingin dibuat berdasarkan datasheet yang sudah diunduh pada langkah sebelumnya. Pada langkah ini RNN akan dipancing untuk memprediksi kataselanjutnya berdasarkan seed textyang sudah ditentukan sebelumnya.

In [ ]:
def predict_next_words(seed_text, next_words):
  for _ in range(next_words):
    token_list = tokenizer.texts_to_sequences([seed_text])[0]
    token_list = pad_sequences([token_list], maxlen=max_sequence_len-1, padding='pre')
    predict_x = model.predict(token_list)
    predicted = np.argmax(predict_x,axis=1)
    #predicted model.predict_classes(token list, verbose-8)
    output_word = ""
    for word, index in tokenizer.word_index.items():
      if index == predicted:
        output_word = word
        break
    seed_text += " " + output_word

  print(seed_text)
  return seed_text

### **Penjelasan Logika & Output: 6. Membuat Perintah Untuk 100 Kata Selanjutnya**

**Logika Kode:**
Ini definisi fungsi `predict_next_words` buat nge-tes model kita. Fungsi ini bakal nge-generasi teks baru.
1.  `for _ in range(next_words):`: Nge-*loop* sebanyak `next_words` (misal 100 kali).
2.  `token_list = tokenizer.texts_to_sequences([seed_text])[0]`: Ngubah `seed_text` (teks awal) jadi urutan angka.
3.  `token_list = pad_sequences(...)`: Nyiapin inputnya. Di-*padding* di depan biar panjangnya pas (10 token) kayak data `xs`.
4.  `predict_x = model.predict(token_list)`: Ini dia prediksinya. Model ngasih *output* probabilitas buat 2900 kata.
5.  `predicted = np.argmax(predict_x,axis=1)`: Milih kata dengan probabilitas tertinggi. `np.argmax` ngambil indeks (angka token) dari kata yang paling mungkin.
6.  `for word, index in tokenizer.word_index.items():`: Nge-*loop* kamus.
7.  `if index == predicted:`: Kalo indeksnya cocok sama hasil prediksi, kita nemu katanya.
8.  `output_word = word`: Simpen katanya.
9.  `seed_text += " " + output_word`: Nambahin kata yang baru ditebak ke `seed_text`.
10. *Loop* tadi ngulang lagi, tapi sekarang `seed_text`-nya udah lebih panjang satu kata. Modelnya jadi nebak kata berikutnya berdasarkan teks yang baru ditambahin.
11. `print(seed_text)`: Setelah *loop* 100 kali selesai, *print* hasil akhirnya.

**Output/Efek:**
Gak ada *output* pas sel ini dijalanin, karena ini cuma definisi fungsi.

# 7. Output Artikel dengan Seed Text

In [ ]:
seed_text = '1THREEPIO'
next_words = 100

generated_text = predict_next_words(seed_text, next_words)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━

### **Penjelasan Logika & Output: 6. Membuat Perintah Untuk 100 Kata Selanjutnya**

**Logika Kode:**
Ini definisi fungsi `predict_next_words` buat nge-tes model kita. Fungsi ini bakal nge-generasi teks baru.
1.  `for _ in range(next_words):`: Nge-*loop* sebanyak `next_words` (misal 100 kali).
2.  `token_list = tokenizer.texts_to_sequences([seed_text])[0]`: Ngubah `seed_text` (teks awal) jadi urutan angka.
3.  `token_list = pad_sequences(...)`: Nyiapin inputnya. Di-*padding* di depan biar panjangnya pas (10 token) kayak data `xs`.
4.  `predict_x = model.predict(token_list)`: Ini dia prediksinya. Model ngasih *output* probabilitas buat 2900 kata.
5.  `predicted = np.argmax(predict_x,axis=1)`: Milih kata dengan probabilitas tertinggi. `np.argmax` ngambil indeks (angka token) dari kata yang paling mungkin.
6.  `for word, index in tokenizer.word_index.items():`: Nge-*loop* kamus.
7.  `if index == predicted:`: Kalo indeksnya cocok sama hasil prediksi, kita nemu katanya.
8.  `output_word = word`: Simpen katanya.
9.  `seed_text += " " + output_word`: Nambahin kata yang baru ditebak ke `seed_text`.
10. *Loop* tadi ngulang lagi, tapi sekarang `seed_text`-nya udah lebih panjang satu kata. Modelnya jadi nebak kata berikutnya berdasarkan teks yang baru ditambahin.
11. `print(seed_text)`: Setelah *loop* 100 kali selesai, *print* hasil akhirnya.

**Output/Efek:**
Gak ada *output* pas sel ini dijalanin, karena ini cuma definisi fungsi.

In [ ]:
seed_text = 'why lovest thou that which not gladly'
generated_text = predict_next_words(seed_text, next_words)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 66ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━

### **Penjelasan Logika & Output: 7. Output Artikel dengan Seed Text**

**Logika Kode:**
Ini bagian ngejalanin fungsi prediksi yang udah kita buat.
1.  `seed_text = '1THREEPIO'`: Nentuin teks awal buat mancing model. *Catatan: '1THREEPIO' ini kayaknya kata yang gak ada di dataset Shakespeare, jadi hasilnya bakal aneh.*
2.  `next_words = 100`: Nyuruh model buat nge-generasi 100 kata baru.
3.  `generated_text = predict_next_words(...)`: Manggil fungsinya.
4.  Sel berikutnya, dicoba lagi pake *seed text* yang lebih masuk akal: `seed_text = 'why lovest thou that which not gladly'`.

**Output/Efek:**
* Pertama, dia bakal nge-print *log* `model.predict` 100 kali (karena di dalem *loop*).
* Terus, dia bakal nge-print teks hasil generasinya.
* Buat *seed* `'1THREEPIO'`, hasilnya: `1THREEPIO to the love of the love of be show thee...` Ini keliatan modelnya "macet" (*looping*), nebak kata 'thee' dan 'i' terus-terusan. Ini wajar karena *seed text*-nya aneh dan *training*-nya baru 10 *epoch*.
* Buat *seed* `'why lovest thou...'`, hasilnya: `why lovest thou that which not gladly thee in thee love the love of show thee...` Hasilnya masih sama-sama "macet", nunjukkin modelnya masih perlu *training* jauh lebih lama (lebih dari 100 *epoch*) biar bisa ngasih hasil yang lebih variatif.